crates/update/src/cli.rs

```diff
@@ -0,0 +1,138 @@
+#![allow(clippy::disallowed_macros)]
+
+use std::ffi::OsString;
+use std::future::Future;
+use std::process::ExitCode;
+
+use anyhow::Context;
+use spacetimedb_paths::{RootDir, SpacetimePaths};
+
+mod install;
+mod link;
+mod list;
+mod uninstall;
+mod upgrade;
+mod r#use;
+
+/// Manage installed spacetime versions
+#[derive(clap::Parser)]
+#[command(bin_name = "spacetime")]
+pub struct Args {
+    #[arg(long)]
+    root_dir: Option<RootDir>,
+    #[command(subcommand)]
+    cmd: Subcommand,
+}
+
+impl Args {
+    pub fn exec(self) -> anyhow::Result<ExitCode> {
+        let paths = match &self.root_dir {
+            Some(root_dir) => SpacetimePaths::from_root_dir(root_dir),
+            None => SpacetimePaths::platform_defaults()?,
+        };
+        match self.cmd {
+            Subcommand::Cli { args: mut cli_args } => {
+                if let Some(root_dir) = &self.root_dir {
+                    cli_args.insert(0, OsString::from_iter(["--root-dir=".as_ref(), root_dir.as_ref()]));
+                }
+                crate::proxy::run_cli(Some(&paths), None, cli_args)
+            }
+            Subcommand::Version(version) => version.exec(&paths).map(|()| ExitCode::SUCCESS),
+            Subcommand::SelfInstall { install_latest } => {
+                let current_exe = std::env::current_exe().context("could not get current exe")?;
+                let suppress_eexists = |r: std::io::Result<()>| {
+                    r.or_else(|e| (e.kind() == std::io::ErrorKind::AlreadyExists).then_some(()).ok_or(e))
+                };
+                suppress_eexists(paths.cli_bin_dir.create()).context("could not create bin dir")?;
+                suppress_eexists(paths.cli_config_dir.create()).context("could not create config dir")?;
+                suppress_eexists(paths.data_dir.create()).context("could not create data dir")?;
+                paths
+                    .cli_bin_file
+                    .create_parent()
+                    .and_then(|()| std::fs::copy(&current_exe, &paths.cli_bin_file))
+                    .context("could not install binary")?;
+
+                if install_latest {
+                    upgrade::Upgrade {}.exec(&paths)?;
+                }
+
+                Ok(ExitCode::SUCCESS)
+            }
+        }
+    }
+}
+
+#[derive(clap::Subcommand)]
+enum Subcommand {
+    Version(Version),
+    #[command(hide = true)]
+    Cli {
+        #[clap(allow_hyphen_values = true)]
+        args: Vec<OsString>,
+    },
+    SelfInstall {
```

Comment:
I think this subcommand is new since my last review? What motivated this?

Reply:
This is for the install script, that we can let spacetimedb-update handle the directory structure stuff instead of the bash or powershell needing to. This is how rustup does it.